[SPARK-48031][SQL] Support view schema evolution

srielau opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Add the following syntax to CREATE VIEW... WITH SCHEMA ...:
```
CREATE [ OR REPLACE ] [ TEMPORARY ] VIEW [ IF NOT EXISTS ] view_name
    [ column_list ]
    [ schema_binding ] 
    [ COMMENT view_comment ]
    [ TBLPROPERTIES clause ]
    AS query

schema_binding
   WITH SCHEMA { BINDING | [TYPE] EVOLUTION | COMPENSATION }

column_list
   ( { column_alias [ COMMENT column_comment ] } [, ...] )
```

Allow changing of the schema binding in:
```
ALTER VIEW view_name  schema_binding
```

The semantic is:

Schema_binding
Optionally specifies how the view adapts to changes to the schema of the query due to changes in the underlying object definitions.
This clause is not supported for temporary views.

- BINDING
The view will become invalid if the query column-list changes except for the following conditions:
The column-list includes a star clause, and there are additional columns. These additional columns are ignored.
The type of one or more columns changed in a way that allows them to be safely cast to the types using implicit casting rules.

- COMPENSATION
The view will become invalid if the query column list changes except for the following conditions:
The column-list includes a star clause, and there are additional columns. These additional columns are ignored.
The type of one or more columns changed in a way that allows them to be cast using explicit cast rules.
This is the default behavior.

- TYPE EVOLUTION
The view will adopt any changes to types in the query column list into is own definition when such a change is detected upon reference of the view.

- EVOLUTION
Behaves like TYPE EVOLUTION and also adopts changes in column names or added and dropped columns if the view does not include an explicit column list.
The view will only be invalidated if the query cannot be parsed anymore, or the optional view column_list does not match the number of expressions in the query select-list anymore.


We also introduce a new SQL Config:
spark.sql.defaultViewSchemaBinding
 Control the default behavior of views when the underlying schema changes.
 Valid values are: 
* COMPENSATION - Any supported casts.
* DISABLED - Disable the feature.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Schema changes are a frequent occurrence, especially when ingesting data.
In the course of it most frequently:

- Types may need to be widened
- Columns are added to tables
- Fields are added to structs

The traditional SCHEMA BINDING behavior makes schema evolution very hard since it invalidates views agressively.
This causes erros, requiring user intervention.
Allowing views to be created to "roll with the punches" or tolerate changes in the underlying schema improves uptime.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, this is a new feature with new grammar and a new config to influence its default.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
New tests are added

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

srielau committed:
[SPARK-48031] Views with SCHEMA EVOLUTION

srielau commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -1323,7 +1323,13 @@
+    "message" : [
+      "The feature <featureName> is not enable. Consider setting the config: <configKey> to <configValue>. To enable this capability."
+    ],
+    "sqlState": "56038"
```
```suggestion
    "sqlState" : "56038"
```

srielau commented:
@cloud-fan @gengliangwang This is ready for an initial review.
Other than docs, the main open issue is NOT NULL behavior and COMMENTs.

nchammas commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -1325,6 +1325,12 @@
   },
+  "FEATURE_NOT_ENABLED" : {
+    "message" : [
+      "The feature <featureName> is not enable. Consider setting the config: <configKey> to <configValue>. To enable this capability."
```
Typo: "not [enabled]"

Typo: "the config `<configKey>` to `<configValue>` to enable this" (remove `:` and `.`)

nchammas commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -4481,6 +4487,11 @@
       },
+      "TEMPORARY_VIEW_WITH_SCHEMA_BINDING_MODE" : {
+        "message" : [
+          "Temporary views cannot be created with the WITH SCHEMA clause. Recreate the temporary view when teh underlying schema changes, or use a persisted view."
```
Typo: "when [the] underlying"

cloud-fan replied:
```suggestion
          "Temporary views cannot be created with the WITH SCHEMA clause. Recreate the temporary view when the underlying schema changes, or use a persisted view."
```

srielau commented:
@cloud-fan @gengliangwang This is ready now. Please review.
Docs are included in this PR.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1700,6 +1700,21 @@ object SQLConf {
+  val VIEW_SCHEMA_BINDING = buildConf("spark.sql.defaultViewSchemaBinding")
+    .doc("Control the default behavior of views when the underlying schema changes. " +
+      "Valid values are: " +
+      "'BINDING': Safe casts only. " +
```
probably we need to explain the difference between the followings:
* Safe casts
* supported casts
* Adopt type changes

gengliangwang replied:
Also, 5 modes seem hard to use..Do we consider reducing it to 3?

srielau replied:
Reducing it to three is hard:
* UNSUPPORTED we need to switch of the feature, including dumping at DESCRIBE EXTENDED, and not storing anything fr new views.
* BINDING is the current default
* COMPENSATION is the new default we have been drifting towards (see changes to DECIMAL (done), and STRUCT (proposed)).
* That only leaves TYPE EVOLUTION vs EVOLUTION.
Slim pickings...

What we could do is, support only the first three for config.
So TYPE EVOLUTION and EVOLUTION is one you must state explicitly.
We could be even more aggressive and say: UNSUPPORTED and COMPENSATION are the only options in config.

Now, to the docs. The description on CREATE VIEW is more verbose. I could point there
The way we (don't) document configs really discourages being verbose. After all SET -v is pretty much it.

srielau replied:
@gengliangwang @cloud-fan 
I have reduced the config to DISABLED/COMPENSATION. That makes it de-factor a feature flag with the option to extend it to a session level default config of the need arises.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1700,6 +1700,21 @@ object SQLConf {
+    .version("4.0.0")
+    .stringConf
+    .transform(_.toUpperCase(Locale.ROOT))
+    .checkValues(Set("BINDING", "COMPENSATION", "TYPE EVOLUTION", "EVOLUTION", "UNSUPPORTED"))
```
Let's rename `TYPE EVOLUTION` to `TYPE_EVOLUTION`, so that it is consistent with other values such as `CodegenObjectFactoryMode`

srielau replied:
If we do this, would we also need to change it in the grammar for CREATE VIEW?

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -401,6 +401,25 @@ case class CatalogTable(
+    if (SQLConf.get.viewSchemaBinding == "UNSUPPORTED") {
+      SchemaUnsupported
+    } else {
+      val schemaMode = properties.getOrElse(VIEW_SCHEMA_MODE, SchemaCompensation.toString)
```
Shall we use `SQLConf.get.viewSchemaBinding` as the default value instead of `SchemaCompensation.toString`? If we change the default value of the SQL conf in the future, we may miss updating this one.

srielau replied:
I was considering that, but the case here is a view which predates the feature.
So if we pick up the config we would have views which behavior changes based on a session config.
That seems weird, even more so if they can refresh themselves.
I'd rather keep the config to only affect CREATE  VIEW.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -401,6 +401,25 @@ case class CatalogTable(
+        case SchemaBinding.toString => SchemaBinding
+        case SchemaEvolution.toString => SchemaEvolution
+        case SchemaTypeEvolution.toString => SchemaTypeEvolution
+        case SchemaCompensation.toString => SchemaCompensation
```
nit: let's handle the default case in the match here.

srielau replied:
Not sure what you want me to do here.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -491,6 +510,9 @@ case class CatalogTable(
     if (tableType == CatalogTableType.VIEW) {
       viewText.foreach(map.put("View Text", _))
       viewOriginalText.foreach(map.put("View Original Text", _))
+      if (SQLConf.get.viewSchemaBinding != "UNSUPPORTED") {
```
Shall we read from the properties instead of getting it from the SQLConf?

srielau replied:
That would not allow us to switch the feature off for DESCRIBE.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -4996,6 +4996,60 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
+        case SchemaTypeEvolution.toString => SchemaTypeEvolution
+        case other => SchemaUnsupported
+      }
+    } else if (SQLConf.get.viewSchemaBinding == "UNSUPPORTED") {
```
```suggestion
    } else if (conf.viewSchemaBinding == "UNSUPPORTED") {
```

gengliangwang commented:
> BINDING
The view will become invalid if the query column-list changes except for the following conditions:
The column-list includes a star clause, and there are additional columns. These additional columns are ignored.
The type of one or more columns changed in a way that allows them to be safely cast to the types using implicit casting rules.
This is the default behavior.


@srielau The PR description seems wrong. The default is `COMPENSATION`, right?

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -1304,7 +1315,8 @@ case class CreateView(
     allowExisting: Boolean,
-    replace: Boolean) extends BinaryCommand with CTEInChildren {
+    replace: Boolean,
+    viewSchemaMode: ViewSchemaMode) extends BinaryCommand with CTEInChildren {
```
Why do we need to show `viewSchemaMode` in creating view? Shall we just show it for Alter View?

srielau replied:
We need this

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1700,6 +1700,16 @@ object SQLConf {
+    .version("4.0.0")
+    .stringConf
+    .transform(_.toUpperCase(Locale.ROOT))
+    .checkValues(Set("COMPENSATION", "DISABLED"))
```
why don't we allow other modes to be the default?

srielau replied:
We did, @gengliangwang raised the question whether that is needlessly complex.
I have been splitting the baby by removing all but these two. having on/off config which is not a boolean. 
If users demand it we can easily expand the config to more values, if not we can deprecate the the config and remove it.

gengliangwang replied:
Shall we also support only `COMPENSATION` in the `WITH SCHEMA` clause? We should make them consistent.

gengliangwang replied:
If you prefer having 5 modes in the WITH SCHEMA clause, let's have 5 modes in the SQL Conf as well. Sorry for causing the misunderstanding.

cloud-fan commented on docs/sql-ref-syntax-ddl-create-view.md:
```diff
@@ -61,6 +61,17 @@ CREATE [ OR REPLACE ] [ [ GLOBAL ] TEMPORARY ] VIEW [ IF NOT EXISTS ] view_ident
     * `[ TBLPROPERTIES ( property_name = property_value [ , ... ] ) ]` to add metadata key-value pairs.
+    * `[ WITH SCHEMA { BINDING | COMPENSATION | [ TYPE ] EVOLUTION } ]` to specify how the view reacts to schema changes
+
+      This clause is not supported for `TEMPORARY` views.
```
Temp view and persistent view share the same view resolution code path, what should be the mode for temp view? Still BINDING as before?

srielau replied:
Temporary views are governed by the config. If the feature is disabled its unchanged (BINDING ). Otherwise COMPENSATION (our new default).

cloud-fan commented on docs/sql-ref-syntax-ddl-create-view.md:
```diff
@@ -80,6 +91,10 @@ CREATE GLOBAL TEMPORARY VIEW IF NOT EXISTS subscribed_movies
         FROM movies AS mo INNER JOIN members AS mb 
         ON mo.member_id = mb.id;
+
+-- Create a view filtering the `orders` table which will adjustto schema changes in `orders`.
```
```suggestion
-- Create a view filtering the `orders` table which will adjust to schema changes in `orders`.
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala:
```diff
@@ -35,6 +35,56 @@ object EliminateView extends Rule[LogicalPlan] with CastSupport {
+ * replace a view in [[CreateViewStatement]].
+ */
+sealed trait ViewSchemaMode {
+  override val toString: String = getClass.getSimpleName.stripSuffix("$")
```
We don't need this override since all subclasses of `ViewSchemaMode` overrides `toString`

srielau replied:
Intelij seems to disagree
<img width="553" alt="Screenshot 2024-05-09 at 9 13 37 AM" src="https://github.com/apache/spark/assets/3514644/4af555f4-742d-4a86-a605-02ba14ee129c">

cloud-fan replied:
I mean remove this entire line, not just the keyword `override`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1700,6 +1700,16 @@ object SQLConf {
+  val VIEW_SCHEMA_BINDING_MODE = buildConf("spark.sql.viewSchemaBindingMode")
+    .doc("Set to DISABLE to disable the WITH SCHEMA clause for view DDL and suppress the line in " +
+      " DESCRIBE EXTENDED. The default, and only other value, is COMPENSATION. Views without " +
+      " WITH SCHEMA clause are defaulted to WITH SCHEMA COMPENSATION.")
```
shall we have different defaults for:
1. existing views
2. newly created views without the `WITH SCHEMA` clause.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -546,6 +551,13 @@ class SparkSqlAstBuilder extends AstBuilder {
+      val schemaBinding = visitSchemaBinding(ctx.schemaBinding(0))
+      val finalSchemaBinding =
+        if (schemaBinding == SchemaEvolution && userSpecifiedColumns.nonEmpty) {
+          SchemaTypeEvolution
```
shall we fail in this case?

srielau replied:
I have been going back and forth on this.
If we returned an error, that would preclude a future config default of EVOLUTION, as it would rquire all views with column list to use an explicit WITH TYPE EVOLUTION.
If we drive this further on might also argue that we should raise an error on EVOLUTION without SELECT * in it.
Basically, without SELECT *: EVOLUTION ==  TYPE EVOLUTION.
Even without all this, I think not returning an error is "nicer", it's not that there are many choices and we need the users help o decide what they want us to do.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:
```diff
@@ -318,6 +329,51 @@ case class AlterViewAsCommand(
+  private def alterPermanentView(session: SparkSession, viewSchemaMode: ViewSchemaMode): Unit = {
+    val viewMeta = session.sessionState.catalog.getTableMetadata(name)
+
+    // Detect cyclic view reference on ALTER VIEW.
```
we don't need to do this as the view query is not changed.

srielau replied:
Ah, I had removed the code, and left the comment

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:
```diff
@@ -620,3 +621,66 @@ object CollationCheck extends (LogicalPlan => Unit) {
+object SyncViewsCheck extends (LogicalPlan => Unit) {
+    def apply(plan: LogicalPlan): Unit = {
+    plan.foreach {
+      case View(metaData, isTempView, viewQuery)
+        if (metaData.viewSchemaMode != SchemaUnsupported && !isTempView) =>
```
```suggestion
      case View(metaData, isTempView, viewQuery)
        if (metaData.viewSchemaMode != SchemaUnsupported && !isTempView) =>
```
```suggestion
      case View(metaData, false, viewQuery)
        if metaData.viewSchemaMode == SchemaTypeEvolution || metaData.viewSchemaMode == SchemaEvolution =>
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:
```diff
@@ -620,3 +621,66 @@ object CollationCheck extends (LogicalPlan => Unit) {
+        val viewSchemaMode = metaData.viewSchemaMode
+        if (viewSchemaMode == SchemaTypeEvolution || viewSchemaMode == SchemaEvolution) {
+          val viewFields = metaData.schema.fields
+          val planFields = plan.schema.fields
```
why do we care `plan.schema`? It's an arbitrary plan that reads a view.

srielau replied:
I've replaced plan with viewQuery.
I think these are the same though, right? view() just passes the schema through

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/view-schema-binding.sql:
```diff
@@ -0,0 +1,429 @@
+-- This test suits check the WITH SCHEMA clause
+-- Make sure we use ANSI_MODE to catch casting errors
```
shall we always use ansi cast regardless of the ansi config?

cloud-fan commented on sql/core/src/test/resources/sql-tests/results/view-schema-binding.sql.out:
```diff
@@ -0,0 +1,3081 @@
+
+
+-- !query
+SET spark.sql.viewSchemaBindingMode
```
why do we check the config value here? We already checked the default value at the beginning.

srielau replied:
Will rework the QA.

srielau commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:
```diff
@@ -437,6 +493,21 @@ object ViewHelper extends SQLConfHelper with Logging {
+    if (viewSchemaMode == SchemaUnsupported) {
+      Map.empty
+    } else {
+      val modeNameJson = JString(viewSchemaMode.toString)
```
Cargo cult...

srielau commented:
@cloud-fan @gengliangwang I have addressed all comments (except the "override" on the, which intelij and Wenchen appear to be of different opinions).
The testsuite has been split up into:
* Default behavior depending on config.
* 1 file per mode.

I don't see a good way to have one file with 4 distinct modes since they have sufficient  syntax and test differences that I think "spaghetti" is more readable than trying to maximize shared text.

gengliangwang commented on sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala:
```diff
@@ -224,6 +224,7 @@ abstract class BaseSessionStateBuilder(
         TableCapabilityCheck +:
         CommandCheck +:
         CollationCheck +:
+        SyncViewsCheck +:
```
The check rules should not come with side effects. Ideally it should check the whole plan and throw exception if something is wrong. However, it seems that there is side effect in this new rule:
```
session.sessionState.catalog.alterTable(updatedViewMeta)
```

cloud-fan replied:
Unfortunately we don't have an analyzer extension point to run at the very end of the analysis phase. We can add one later, but for now `CheckAnlysis` is the best place to do it. Mabybe we can rename this rule to indicate that it has side effects?

srielau replied:
"ViewSyncSchemaToMetaStore" coming

srielau commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -945,54 +945,73 @@ class SessionCatalog(
-      metadata.schema.zipWithIndex.map { case (field, index) =>
-        val col = GetColumnByOrdinal(index, field.dataType)
-        Alias(UpCast(col, field.dataType), field.name)(explicitMetadata = Some(field.metadata))
+        // For view created by hive, the parsed view plan may have different output columns with
```
Good catch! I didn't understand the code and somehow concluded it doesn't apply for Hive tables.

srielau replied:
I've replicated the code above

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala:
```diff
@@ -35,6 +35,56 @@ object EliminateView extends Rule[LogicalPlan] with CastSupport {
+ * It can tolerate extra columns in SELECT * and upcast to more generic types.
+ */
+object SchemaBinding extends ViewSchemaMode {
+  override val toString: String = "BINDING"
```
```suggestion
  override def toString: String = "BINDING"
```
since the function body is just a string literal

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -945,54 +945,87 @@ class SessionCatalog(
+            case SchemaTypeEvolution => col
+            case other => throw SparkException.internalError("Unexpected ViewSchemaMode")
+          }
+          Alias(cast, field.name)(explicitMetadata = Some(field.metadata))
```
nit: let's create a methond to do the cast, to avoid code duplication

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:
```diff
@@ -620,3 +621,65 @@ object CollationCheck extends (LogicalPlan => Unit) {
+        val viewQueryFields = viewQuery.schema.fields
+        val session = SparkSession.getActiveSession.get
+        val redoSignature =
+          viewSchemaMode == SchemaEvolution && viewFields.length != viewQueryFields.length
```
what happens if the number of columns doesn't change but the column names are all different? shall we still redo signature?

srielau replied:
Yep, I had the test, but didn't catch the bad describe info.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:
```diff
@@ -620,3 +621,65 @@ object CollationCheck extends (LogicalPlan => Unit) {
+          val planField = viewQueryFields(index)
+          (field.dataType != planField.dataType ||
+            field.nullable != planField.nullable ||
+            (viewSchemaMode == SchemaEvolution && (
```
if we hit this condition, shall we also redo signature?

srielau replied:
Yes, If we need to refresh the view for any reason and we are in EVOLUTION we need to regen.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala:
```diff
@@ -224,6 +224,7 @@ abstract class BaseSessionStateBuilder(
         TableCapabilityCheck +:
         CommandCheck +:
         CollationCheck +:
+        ViewSyncSchemaToMetaStore +:
```
we should put this rule in `HiveSessionStateBuilder` as well

srielau replied:
There already (?)

cloud-fan commented on sql/core/src/test/resources/sql-tests/results/view-schema-binding-config.sql.out:
```diff
@@ -0,0 +1,794 @@
+
+
+-- !query
+SHOW TABLE EXTENDED LIKE 'v'
```
what extra information we want to get from `SHOW TABLE EXTENDED`?

cloud-fan commented on sql/core/src/test/resources/sql-tests/results/view-schema-binding.sql.out:
```diff
@@ -0,0 +1,407 @@
+-- Automatically generated by SQLQueryTestSuite
+-- !query
+SET spark.sql.ansi.enabled = true
```
do we need to? It shouldn't affect any behavior here.

srielau commented:
@cloud-fan @gengliangwang All of @cloud-fan comments have been accepted and addressed.
Shall we merge it?

cloud-fan commented:
the k8s test failure is unrelated, thanks, merging to master!

HyukjinKwon commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala:
```diff
@@ -742,7 +742,7 @@ abstract class SessionCatalogSuite extends AnalysisTest with Eventually {
     import org.apache.spark.sql.catalyst.dsl.expressions._
     val projectList = metadata.schema.map { field =>
-      UpCast(
+      Cast(
```
Seems like this breaks non-ANSI build. Made a followup: https://github.com/apache/spark/pull/46614
